source/shared/ed25519-lc.c

```diff
@@ -0,0 +1,351 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/cal/ed25519.h>
+#include <aws/cal/private/opensslcrypto_common.h>
+
+#include <aws/common/device_random.h>
+#include <aws/common/encoding.h>
+#include <aws/common/ref_count.h>
+
+#include <openssl/evp.h>
+
+struct aws_ed25519_key_pair {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+
+    EVP_PKEY *key;
+};
+
+static void s_ed25519_destroy_key(void *key_pair) {
+    if (key_pair == NULL) {
+        return;
+    }
+
+    struct aws_ed25519_key_pair *lc_key_pair = (struct aws_ed25519_key_pair *)(key_pair);
+
+    if (lc_key_pair->key != NULL) {
+        EVP_PKEY_free(lc_key_pair->key);
+    }
+
+    aws_mem_release(lc_key_pair->allocator, lc_key_pair);
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_new_generate(struct aws_allocator *allocator) {
+#if defined(OPENSSL_IS_OPENSSL) && OPENSSL_VERSION_NUMBER <= 0x10101000L
+    /* ed25519 support does not exist prior to 1.1.1 */
+    aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    return NULL;
+#else
+    EVP_PKEY *pkey = NULL;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);
+    if (ctx == NULL) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        return NULL;
+    }
+
+    if (EVP_PKEY_keygen_init(ctx) <= 0) {
+        goto on_error;
+    }
+
+    if (EVP_PKEY_keygen(ctx, &pkey) <= 0) {
+        goto on_error;
+    }
+
+    struct aws_ed25519_key_pair *key_pair = aws_mem_calloc(allocator, 1, sizeof(struct aws_ed25519_key_pair));
+
+    aws_ref_count_init(&key_pair->ref_count, key_pair, s_ed25519_destroy_key);
+    key_pair->allocator = allocator;
+    key_pair->key = pkey;
+
+    return key_pair;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    return NULL;
+#endif
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_acquire(struct aws_ed25519_key_pair *key_pair) {
+    return aws_ref_count_acquire(&key_pair->ref_count);
+}
+
+struct aws_ed25519_key_pair *aws_ed25519_key_pair_release(struct aws_ed25519_key_pair *key_pair) {
+    if (key_pair != NULL) {
+        aws_ref_count_release(&key_pair->ref_count);
+    }
+    return NULL;
+}
+
+static struct aws_byte_cursor s_key_type_literal = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("ssh-ed25519");
+
+int s_ed25519_openssh_encode_public_key(const struct aws_ed25519_key_pair *key_pair, struct aws_byte_buf *out) {
+    if (!aws_byte_buf_write_be32(out, s_key_type_literal.len) ||
+        aws_byte_buf_append(out, &s_key_type_literal) != AWS_OP_SUCCESS || !aws_byte_buf_write_be32(out, 32)) {
```

Comment:
If a function fails, the length of the buffer it writes to should remain unchanged.

But this function can fail midway.

Either have this goto error and put the length back where it was, or have the public functions that use these helpers do that if any of the helpers fail